Use memory cache to store pages and avoid keeping them in the DOM

f-elix opened the pull request:
This PR is making a change to the way the client-side routing is currently working.

Instead of keeping already visited pages in the DOM and hiding them with CSS, they are cached in an object, using the page `key` as the object keys, and the DOM node as the value. This way, we can safely remove the page we're leaving from the DOM and re-append it from the cache if it is re-visited.

I think this is a better approach overall. I tested these changes and everything seems to be working, but I am not familiar with the entirety of the framework and it is very possible that I missed some things.

f-elix committed:
Use memory cache to store pages and avoid keeping them in the DOM

fhamon commented on src/app-pages.js:
```diff
@@ -12,10 +12,11 @@
 	const pageModels = {};
 	const pageInstances = {};
 	const activeRoutes = {};
+	const loadedPages = {};
```
The loadedPages should be in their own page instance.

fhamon replied:
We also need a way to gc on unused pages.

f-elix replied:
1) Not sure I understand what you mean by that. Each page instance should have a `loadedPages` cache object?

2) How do we determine if a page is unused?

fhamon replied:
1. each loaded page has a framework page instance defined in `pageInstances` the loaded dom should be attached to the instance.

2. it's not the current url. and is only an opt-in

f-elix committed:
Use pageInstance to cache page node

f-elix committed:
Remove comment

nitriques commented:
I really like the idea, but there is a couple of things that bothers me, in no particular orders.

1. We are removing the immutability of the page object, which can creates side effects.
2. We are keeping a DOMNode in memory, which _may_ create memory leaks.
3. It is removing the node too soon (i.e. before we have added the other one).
4. It prevents us from showing 2 pages at the same time (background page for a modal page).

That being said, I am not against it. But we need to address all points, either by fixing them or by accepting the bracking changes.

f-elix commented:
> 1. We are removing the immutability of the page object, which can creates side effects.
> 2. We are keeping a DOMNode in memory, which _may_ create memory leaks.
> 3. It is removing the node too soon (i.e. before we have added the other one).
> 4. It prevents us from showing 2 pages at the same time (background page for a modal page).

1. Wasn't already mutable with `setInited()`? If not, maybe I misunderstand.
2. I was wondering about that. We are currently keeping the node in the DOM, is there a performance/memory management advantage to that? How could it create a memory leak and how could we prevent it?
3. I didn't touch the order of the `enterLeave()` function in `app-mediator.js`. Maybe we could add a step where we append the nextPage hidden, remove the previousPage and then show the nextPage. 
4. Is this currently supported? I'm not sure how I would do that with the current framework version.

nitriques commented:
@f-elix 

> 1. Wasn't it already mutable with setInited()? If not, maybe I misunderstand.

In a purely strict sense, maybe not, but it does have the proper qualities we are looking for, i.e. it still prevents side effects.
There is no way for it to become false once it is set to true. There is no parameters in setInited().
Since we can guarantee that setInited will be called, the object will reach full immutability.

> 2. I was wondering about that. We are currently keeping the node in the DOM, is there a performance/memory management advantage to that?

disadvantage: using a bit more memory because all event listeners are still there, images are still loaded, etc...
advantages: there is no 'bootstraping' to be done since we got rid of nothing. We trade memory (more) for cpu (less) to get more responsiveness. We can guarantee that our code won't leak by holding a reference to a dead (removed) element.

And that's, IMHO, the best thing about that way of doing it. But it has serious limit.

> Also, in what way could keeping the html string in js create a memory leak and how could we prevent it?

That was my idea ;) We would still need to do the bootstraping, but we could skip the network layer, when it slow. (i.e. no caching reverse proxy).

> 3. I didn't touch the order of the enterLeave()

Indeed, but the fact that it gets removed has a side effect to what we currently do in enter. I may have overlooked that part, I'll review again tomorrow, which a fresh head.

> 4. Is this currently supported? I'm not sure how I would do that with the current framework version.

Well, since the page is in the DOM, one could revert the display none style from user land code (user land = not in the framework)

f-elix commented:
1) Regarding immutability: 

I can't think of a way of keep the page object immutable with the way I wrote the changes. However, in my first attempt I added a `loadedPages` object to hold the html strings (with page urls as keys). After a review by @fhamon, I realized it made more sense to keep the html in the page instance object itself so I removed it, but maybe that can be the solution after all?

2) Regarding memory vs cpu usage:

If we are not ready to make that trade off (which would be a valid decision), then we should forget about this PR. Although I think caching the html in js has other advantages, especially when it comes to cleaning up listeners on page leave. Is that a sufficient advantage to continue here?

3) Regarding the option to stack multiple pages:

I see. If we wanted for this to still be possible we would probably have to add an API from the framework for doing so, right? Something like a page option to prevent removing it from the DOM. Is it worth it?

f-elix committed:
Change caching and page change strategy

f-elix commented:
I updated the code with some new ideas I had after your comments.

1) Instead of storing the DOM node in memory, we store a DOM string (with `node.outerHTML`). This way, we can completely remove the page's DOM object from memory and prevent potential memory leaks.

2) Renamed page instance's `setNode` function to `cache`. The function now only updates the page instance's `domString` variable if it is still `undefined`. That way, some kind of immutability is preserved. You can only set the page instance's `domString` once, just like you can only set `isInited` to `true` only once.

3) Renamed page instance's `node` function to `loaded`. The only use there was for this function was to verify that the page had been cached, There is no need for the page instance to expose the `domString`, we can therefore just return `true` or `false` to indicate if the page has been loaded and cached or not.

4) Refactored page instance `enter` and `leave` functions. We first check if the page is already in the DOM, if it is, we just call `next` and return (this is usually the initial page). Otherwise, we just replace the root node `innterHTML` with the page's DOM string. The `leave` function now has nothing to do except to call the `next` callback.

5) I tried to think of a way to keep the page in the DOM in case the user wants to show 2 pages at the same time, but I came to the conclusion that this shouldn't be the responsibility of the framework. This can still be done in userland by preventing the url changer from changing the page and changing the url manually from a module.
